### PR TITLE
integration/k8s: Build custom stress image

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
containerstack/alpine-stress is only built for x86_64 and I cannot find
a substitute. Let's build the image ourselves to have a `stress`
container available on all architectures.

Fixes: #3678
Depends-on: github.com/kata-containers/tests#3636
Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

### CODE REVIEW

File: integration/kubernetes/runtimeclass_workloads/stress/Dockerfile
Comment (jodh-intel) [inline integration/kubernetes/runtimeclass_workloads/stress/Dockerfile:4]: Looking at the original `Dockerfile` (https://github.com/containerstack/alpine-stress/blob/master/Dockerfile) shows that `stress` originates from:

- https://people.seas.harvard.edu/~apw/stress.

That's confirmed by the debian package page too:

- https://packages.debian.org/en/buster/stress.

However, the homepage no longer exists (which presumably explains why sid doesn't list it - https://packages.debian.org/en/sid/stress).

I wonder if we want to use `stress-ng` instead?:

- https://packages.ubuntu.com/hirsute/stress-ng

File: integration/kubernetes/init.sh
Comment (MatthieuSarter) [inline integration/kubernetes/init.sh:167]: This script may be used with containerd or CRI-O, so the CRI-O case should also be handled. It might require to have podman installed, as crictl does not allow to import images.

File: integration/kubernetes/runtimeclass_workloads/stress/Dockerfile
Comment (Jakob-Naucke) [reply integration/kubernetes/runtimeclass_workloads/stress/Dockerfile:4]: Hmmm, I did notice that, and I knew that stress-ng existed. There's also quay.io/kinvolk/stress-ng, which is at least also built for arm64, but not for the IBM architectures, so we'd probably still need the custom build. Let me TAL at how much we would have to change.

File: integration/kubernetes/init.sh
Comment (Jakob-Naucke) [reply integration/kubernetes/init.sh:167]: Yes. This is why this PR depends on #3636, which installs Podman and uses it for osbuilder on Fedora-likes. I.e.
- Fedora-likes will have Podman+CRI-O, and the image will be available to Kubernetes right away after line 166.
- Everything else will have Docker+containerd (lines below).
- I don't think Podman+containerd+k8s is used anywhere, but I just figured I might actually add support for it because it's quite simple.
- I think Docker+CRI-O+k8s is not possible without an external registry, so yes, that is one combination we would drop. But that's why the dependency on #3636 exists.

File: integration/kubernetes/runtimeclass_workloads/stress/Dockerfile
Comment (jodh-intel) [reply integration/kubernetes/runtimeclass_workloads/stress/Dockerfile:4]: Thanks @Jakob-Naucke. Hopefully, it should just be a case of changing the `apt` package name as `stress-ng` is tested on all arches fwics.

/cc @liubin and @GabyCT on the idea of switching to a different "stressor" package though.

Comment (Jakob-Naucke) [reply integration/kubernetes/runtimeclass_workloads/stress/Dockerfile:4]: Yes, but I think the flags in the tests we use it in (`integration/kubernetes/k8s-{memory,oom}.bats`) will change.

File: integration/kubernetes/init.sh
Comment (Jakob-Naucke) [reply integration/kubernetes/init.sh:168]: Updated as mentioned.

File: integration/kubernetes/runtimeclass_workloads/stress/Dockerfile
Comment (Jakob-Naucke) [reply integration/kubernetes/runtimeclass_workloads/stress/Dockerfile]: updated... it seems `stress-ng` is actually (and probably intentionally) forwards-comptaible.

Comment (jodh-intel) [reply integration/kubernetes/runtimeclass_workloads/stress/Dockerfile]: @Jakob-Naucke - nice! And thanks @ColinIanKing! :smile:

**Review by jodh-intel — APPROVED** [review]
Thanks @Jakob-Naucke.

lgtm

**Comment by fidencio** [comment]
Howdy!

I'd like to ask you to rebase your PR atop of current main, as #3668 has been merged and it has a fix that unblocks the CI.

Please, do the rebase and feel free to `@fidencio` in case you do not have the powers to run `/test`.

Sorry for the inconvenience and happy hacking!

**Comment by Jakob-Naucke** [comment]
rebase & fix depends tag
/test

**Comment by fidencio** [comment]
I'm adding the `do-not-merge` label till we get `2.2.0-alpha1` release out.
Once we cut the release I'll come back removing the label from the PRs.

**Comment by Jakob-Naucke** [comment]
added transitive PR dependency
/test

**Comment by fidencio** [comment]
I'm really sorry for the disruption we had, but I'm glad to say that 2.2.0-alpha1 release is finally out and we're removing all the `do-not-merge` labels that were added due to this.

Thanks a lot for your understanding and happy hacking!

**Comment by Jakob-Naucke** [comment]
So we are using Docker+CRI-O+k8s e.g. on the Debian CI (so it definitely works on non-Fedora-likes as well) and that's a combination I'd rather want to support. Taking a look at running a local registry.

**Comment by GabyCT** [comment]
@Jakob-Naucke could you please do a rebase of this PR ...some CI fixes were done that I believe is impacting the CI thanks

**Comment by Jakob-Naucke** [comment]
So the previously anticipated workflow of importing the built image with `ctr --namespace=k8s.io` for containerd or Podman with CRI-O is not going to fly, since Docker+CRI-O is both very much possible and even used in our CI (Debian).
I then looked into telling Docker to use the Kubernetes namespace, which would have looked a little like
```sh
export DOCKER_TLS_VERIFY=1
export DOCKER_HOST="$(yq read "$KUBECONFIG" "clusters[0].cluster.server") | sed "s/https/tcp/")"
yq read "$KUBECONFIG" "clusters[0].cluster.certificate-authority-data" | base64 -d > "$HOME/.kube/ca.pem"
export DOCKER_CERT_PATH="$HOME/.kube"
```

and while I could swear that this worked at one point in a VM where I failed to restart the cluster again, maybe it never worked and it merely worked with the minikube that I used for figuring out these variables.
Anyway, this leads to the error:
```
Error response from daemon: forbidden: User "system:anonymous" cannot get path "/v1.24/build"
```

My fight against resolving this has been unsuccessful and unless someone has an idea, since we recently got [quay.io/kata-containers](https://quay.io/organization/kata-containers), I wonder whether we should use that instead.
/cc @fidencio because I would need a bit of guidance there :)

**Comment by Jakob-Naucke** [comment]
Re-requesting review because a lot has changed. This works with Docker and Podman, and is CRI runtime indepedent.
`docker.io/registry/registry` is not perfect due to rate limiting, ideas for alternatives? There is also `quay.io/ibm/registry`, but unlike `docker.io/ibmcom/registry` (which the PR currently uses for ppc64le & s390x), it only builds for s390x.
/test

**Comment by Jakob-Naucke** [comment]
forgot second substitution in exceed memory test
/test

**Comment by Jakob-Naucke** [comment]
So there is _something_ that `stress-ng` does differently from `stress`. They seem to behave similarly in some quick tests, but while the `k8s-memory` test is working, the `k8s-oom` test clearly works with `stress`, but not with `stress-ng` -- the OOMKill does not seem to happen.
(For reference, it works with these changes:)
```diff
diff --git a/integration/kubernetes/runtimeclass_workloads/pod-oom.yaml b/integration/kubernetes/runtimeclass_workloads/pod-oom.yaml
index 3dfcbd88..7b9d414a 100644
--- a/integration/kubernetes/runtimeclass_workloads/pod-oom.yaml
+++ b/integration/kubernetes/runtimeclass_workloads/pod-oom.yaml
@@ -20 +20 @@ spec:
-      args: ["-c", "sleep 2; stress-ng --vm 2 --vm-bytes 400M --timeout 30s"]
+      args: ["-c", "sleep 2; stress --vm 2 --vm-bytes 400M --timeout 30s"]
diff --git a/integration/kubernetes/runtimeclass_workloads/stress-ng/Dockerfile b/integration/kubernetes/runtimeclass_workloads/stress-ng/Dockerfile
index b25b541c..6546f030 100644
--- a/integration/kubernetes/runtimeclass_workloads/stress-ng/Dockerfile
+++ b/integration/kubernetes/runtimeclass_workloads/stress-ng/Dockerfile
@@ -9 +9 @@ RUN apt-get -y update && \
-    apt-get -y install stress-ng
+    apt-get -y install stress
```

Now I'm debating whether I should track the root cause of this (because this is probably a bug in our testing), or just keep using `stress`.

/cc @jodh-intel

**Comment by Jakob-Naucke** [comment]
Okay, this was because `--vm` uses `malloc` in `stress`, but `mmap` in `stress-ng`, and to use `malloc` in `stress-ng`, you need `malloc`, so this seems to be a classical "failing to RTFM", although I couldn't pinpoint why it wouldn't work with `mmap`.
/test

**Comment by ColinIanKing** [comment]
Yep, at the end of the stress-ng manual it states:

"Note  that the stress-ng cpu, io, vm and hdd tests are different imple‐
 mentations of the original stress tests and hence may produce different
 stress  characteristics."

It's hard to do a clean-room reimplementation of the original stress w/o knowing what it really is doing. :-/

**Comment by Jakob-Naucke** [comment]
I can reproduce this: it does not work under CRI-O. Investigating...

**Comment by Jakob-Naucke** [comment]
I have a hard time putting my finger on the difference between containerd and CRI-O here. But @ColinIanKing could you TAL at this (this is not in a container):
```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:          7.6Gi       819Mi       5.9Gi       382Mi       974Mi       6.2Gi
Swap:            0B          0B          0B
$ stress --vm 1 --vm-bytes 16G --timeout 60s
stress: info: [145711] dispatching hogs: 0 cpu, 0 io, 1 vm, 0 hdd
stress: FAIL: [145711] (415) <-- worker 145712 got signal 9
stress: WARN: [145711] (417) now reaping child worker processes
stress: FAIL: [145711] (451) failed run completed in 3s
$ stress-ng --malloc 1 --malloc-bytes 16G --oomable --timeout 60s
stress-ng: info:  [146265] dispatching hogs: 1 malloc
stress-ng: info:  [146265] successful run completed in 60.14s (1 min, 0.14 secs)
```

shouldn't `stress-ng` be OOMed there?

**Comment by ColinIanKing** [comment]
stress-ng will respawn OOM'd stressor processes since it is designed to keep the stress going until the timeout or bogo-ops limits are reached.  The --oomable option will override this default and allow stressors to be OOM'd without a restart.

**Comment by Jakob-Naucke** [comment]
@ColinIanKing I used `--oomable`, check the line

**Comment by ColinIanKing** [comment]
Ah, I overlooked that --oomable.  So the malloc stressor will allocate memory but Linux uses lazy memory allocation hence pages that are not touched won't cause a page to be physically mapped. The stressor is intended to thrash the libc memory allocator rather than eat loads of memory,  but I could add an option to make this stressor touch each page to force them to be physically mapped in.

**Comment by Jakob-Naucke** [comment]
> So the malloc stressor will allocate memory but Linux uses lazy memory allocation hence pages that are not touched won't cause a page to be physically mapped. The stressor is intended to thrash the libc memory allocator rather than eat loads of memory, but I could add an option to make this stressor touch each page to force them to be physically mapped in.

Ah, that makes sense. I think I will switch back to `stress` for the time being. Such an option would be useful, but don't feel `stress`ed to do it :)

**Comment by Jakob-Naucke** [comment]
I didn't actually push towards my fork, sorry for the noise -- /test

**Comment by ColinIanKing** [comment]
--malloc-touch option now added to stress-ng to force all pages to be physically resident on a malloc/realloc.

**Comment by Jakob-Naucke** [comment]
on ARM/s390x:
```
15:39:54 Err:1 http://ports.ubuntu.com/ubuntu-ports focal InRelease
15:40:14   Temporary failure resolving 'ports.ubuntu.com'
```

Let me check whether there's something wrong with the `libpod` image.
/test

**Comment by Jakob-Naucke** [comment]
that should be fixed, /test

**Comment by Jakob-Naucke** [comment]
all look like one-offs -- /retest-arm /retest-s390x /retest-clh

**Review by GabyCT — APPROVED** [review]

**Comment by fidencio** [comment]
@Jakob-Naucke, out of curiosity, considering that now we have a quay.io account and we have some freedom on managing that, would be worth it to maintain the image there?

**Comment by Jakob-Naucke** [comment]
@jodh-intel I am going to dismiss your review because the PR was very different at the time, it would be dishonest to the processes to merge right now.

> @Jakob-Naucke, out of curiosity, considering that now we have a quay.io account and we have some freedom on managing that, would be worth it to maintain the image there?

@fidencio just now I learnt about [quay.io being able to build based on Dockerfiles](https://docs.quay.io/guides/building.html), but we have yet to figure out the available hardware. We can follow the approach from this PR for the time being.

**Review by jodh-intel — APPROVED** [review]
Thanks @Jakob-Naucke.

lgtm

**Pull request merged**
